service/frontend/workflowHandler.go

```diff
@@ -1620,6 +1623,166 @@ func (wh *WorkflowHandler) GetWorkflowExecutionHistory(
+	sw := wh.startRequestProfile(scope)
+	defer sw.Stop()
+	var err error
+
+	domainID, err := wh.domainCache.GetDomainID(request.GetDomain())
+	if err != nil {
+		return nil, wh.error(err, scope)
+	}
+
+	execution := request.Execution
+	if len(execution.GetWorkflowId()) == 0 {
+		return nil, &gen.BadRequestError{Message: "Invalid WorkflowID."}
+	}
+	// TODO currently, this API is only going to be used by re-send history events
+	// to remote cluster if kafka is lossy again, in the future, this API can be used
+	// by CLI and client, then empty runID (meaning the current workflow) should be allowed
+	if len(execution.GetRunId()) == 0 || uuid.Parse(execution.GetRunId()) == nil {
+		return nil, &gen.BadRequestError{Message: "Invalid RunID."}
+	}
+
+	pageSize := int(request.GetMaximumPageSize())
+	if pageSize < 0 {
+		return nil, &gen.BadRequestError{Message: "Invalid PageSize."}
+	}
+
+	var token *getHistoryContinuationToken
+	// initialize or validate the token
+	// token will be used as a source of truth
+	if request.NextPageToken != nil {
+		token, err = deserializeHistoryToken(request.NextPageToken)
+		if err != nil {
+			return nil, err
+		}
+
+		if execution.GetRunId() != token.RunID ||
+			// during the pagination, branch token should not change
+			bytes.Compare(request.BranchToken, token.BranchToken) != 0 ||
+			// we guarantee to use the first event ID provided in the request
+			request.GetFirstEventId() != token.FirstEventID ||
+			// the next event ID in the request must be <= next event ID from mutable state, when initialized
+			// so as long as customer do not change next event ID during pagination,
+			// next event ID in the token <= next event ID in the request.
+			request.GetNextEventId() < token.NextEventID {
+			return nil, &gen.BadRequestError{Message: "Invalid pagination token."}
+		}
+
+		// for the rest variables in the token, since we do not do hmac,
+		// the only thing can be done is to trust the token:
+		// IsWorkflowRunning: not used
+		// TransientDecision: not used
+		// PersistenceToken: trust
+		// EventStoreVersion: trust
+		// ReplicationInfo: trust
+
+	} else {
+		firstEventID := request.GetFirstEventId()
+		nextEventID := request.GetNextEventId()
+		if firstEventID < 0 || firstEventID > nextEventID {
+			return nil, &gen.BadRequestError{Message: "Invalid FirstEventID && NextEventID combination."}
+		}
+
+		response, err := wh.history.GetMutableState(ctx, &h.GetMutableStateRequest{
+			DomainUUID:          common.StringPtr(domainID),
+			Execution:           execution,
+			ExpectedNextEventId: common.Int64Ptr(common.FirstEventID), // common.FirstEventID means no long poll
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// branch token is provided as part of request, makeing sure that no branching has happened yet
+		// otherwise, caller need to do a reload of history from the begining
+		if request.BranchToken != nil && bytes.Compare(request.BranchToken, response.BranchToken) != 0 {
```

Comment:
I don't think this is necessary.  We always create a new run when branching workflow executions.  So we don't expect multiple branches with the same run id anymore.

Reply:
this is true except that conflict resolution, when that is enabled, a new branch will be created and run ID remains the same.

anyway, removing the branch token for now, since conflict resolution & branching is not implemented.